install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml

```diff
@@ -370,6 +374,19 @@ spec:
 {{- end }}
       hostNetwork: true
       initContainers:
+      # Required to mount cgroup v2 filesystem on the underlying Kubernetes node
+      - name: mount-cgroup
+        env:
+          - name: CGROUP_ROOT
+            value: {{ .Values.cgroupRoot }}
+        command: ['nsenter', '--cgroup=/hostproc/1/ns/cgroup', '--mount=/hostproc/1/ns/mnt', "--",
+                  'sh', '-c',  'mount | grep "$CGROUP_ROOT type cgroup2" || { echo "Mounting cgroup filesystem..."; mount -t cgroup2 none $CGROUP_ROOT; }']
+        image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}{{ if .Values.image.useDigest }}@{{ .Values.image.digest }}{{ end }}"
+        volumeMounts:
+          - mountPath: /hostproc
+            name: host-proc
+        securityContext:
+          privileged: true
```

Comment:
This is a really nasty hack.

Reply:
IMHO saying that this is _a really nasty hack_ doesn't provide much helpful information. Mind elaborating instead?

Reply:
>Bidirectional mount propagation as proposed in the weekly meeting?

@aanm Can you explain how the propagation mount will solve the cgroup namespace problem? Also, It's not guaranteed that cgroup v2 fs is guaranteed to be mounted on the host at `/sys/fs/cgroup` path. 
Example on GKE - 
```
$ mount | grep cgroup2
cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime)
```

Reply:
One think to note is that if we know _where_ the cgroupv2 host fs is mounted in the host, the mount propagation allows us to expose the host cgroupfs inside the agent's container without having to expose /proc [1]. Couldn't we have a configurable helm variable for this location (default /sys/fs/cgroup/) that can be set to a proper value for each platform that deviates (e.g., GKE).

E.g., something like:

```
- hostPath:
       path: {{ .Values.hostCgroupFS }} # /sys/fs/cgroup by default
        type: DirectoryOrCreate
   name: cilium-cgroup
```

and:
```
- mountPath: /var/run/cilium/cgroupv2
   name: cilium-cgroup
```


[1] I've just tested this. I think the way it works is by  doing a bind mount inside the root fs of the container.

Reply:
I think this is may be a fair middle ground. Having said that, I may have a preference for mounting cgroupfs at a cilium specified custom path - 
As far as possible, we should try to keep the internal details as transparent to the end user as possible. It can be error prone to have too many knobs. See this note in the [cgroup manpage](https://man7.org/linux/man-pages/man7/cgroups.7.html) - 

>Note that on many modern systems, systemd(1) automatically mounts
       the cgroup2 filesystem at /sys/fs/cgroup/unified during the boot
       process.

So the cgroupfs mount location may differ across cgroup v1/v2, systemd or not, etc. The default /sys/fs/cgroup path for cgroupfs 2 doesn't necessarily cover a large number of environments.

Reply:
> IMHO saying that this is _a really nasty hack_ doesn't provide much helpful information. Mind elaborating instead?

@brb As mentioned in the review description, bidirectional mounts offered by Kubernetes should help having the behavior that we want without having the mount the entire `/proc`, increasing attack surface, into Cilium's container.

Sure there are environments where the cgroup directory is not the default one but we can hard-code the value of `--cgroup-root` for the environments that we know of. We already have dedicated options for cloud providers so I don't see how much difference it would be having dedicated value for the `--cgroup-root` depending on the cloud provider. We can even mount `/sys/fs`, instead of `/sys/fs/bpf`, inside the cilium-agent container and we could detect if the cgroup is mounted in ` /sys/fs/cgroup/unified` or in ` /sys/fs/cgroup`.

For the more specific environments where users might change the cgroup directory intentionally then we just need to add documentation regarding this in the kube-proxy GSG.

Reply:
>without having the mount the entire /proc, increasing attack surface, into Cilium's container.

We are only mounting `/proc` inside the init container that's short-lived (not inside the agent container). Cilium CLI already [mounts](https://github.com/cilium/cilium-cli/blob/master/install/install.go#L685) `/proc` in order to mount `/sys/fs/bpf`.

Reply:
> > without having the mount the entire /proc, increasing attack surface, into Cilium's container.
> 
> We are only mounting `/proc` inside the init container that's short-lived (not inside the agent container). Cilium CLI already [mounts](https://github.com/cilium/cilium-cli/blob/master/install/install.go#L685) `/proc` in order to mount `/sys/fs/bpf`.

Cilium CLI is still in beta stage, the reason it is mounting proc was to make it easier to deploy Cilium across all environments but that does not mean it's the safest way of mounting bpf fs. We should replace Cilium cli with bidirectional mounting instead of mounting `/proc` in Cilium repo.

Reply:
To reduce the attack surface, wouldn't it be enough to mount `/proc/1/ns/cgroup` and `/proc/1/ns/mnt` instead of `/proc`?

Reply:
Yeah, we could do selective `/proc/*` mounts.

Reply:
`/proc/1/ns/cgrou`p and `/proc/1/ns/cgroup` are symlinks. I've made the change to mount `/proc/1/ns` instead.